src/util/alloc/allocators.rs

```diff
@@ -175,3 +175,33 @@ impl Default for AllocatorSelector {
         AllocatorSelector::None
     }
 }
+
+/// This type describes allocator information. It is used to
+/// generate fast paths for the GC. All offset fields are relative to [Mutator](crate::Mutator).
+#[repr(C, u8)]
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub enum AllocatorInfo {
+    BumpPointer {
+        limit_offset: usize,
+        top_offset: usize,
+    },
+    MarkCompact {
+        limit_offset: usize,
+        top_offset: usize,
+    },
+    Immix {
```

Comment:
I think you can just remove the `Immix` variant as well. Immix's `overflow_alloc` (which uses the large cursor/limit) and `alloc_slow_hot` (which tries to find a recyclable line) is considered as the 'intermediate' path (i.e. between fastpath and slowpath). Normally the compiler don't need to inline it. In that case, generating allocation fastpath for Immix would be the same as `BumpPointer` fastpath (just checking cursor/limit, and adjust cursor).

Reply:
Should not small limit still be checked? Or am I wrong here?

Reply:
You don't have to check "small_limit". If the allocation can fit into the normal buffer (cursor/limit), it will be allocated in the buffer. However, the buffer can be as small as one line (depending on how fragment the heap is). So sometimes the allocation that is larger than one line can't fit into the normal buffer, then the large cursor/limit is used. But that is not in the fastpath path any more.

The actual code that is considered as Immix's fastpath is (which is as simple as a normal bump pointer allocation):
https://github.com/mmtk/mmtk-core/blob/4c1e58a1b293e0e710feebdb7bec876bd98866f8/src/util/alloc/immix_allocator.rs#L75-L78
and
https://github.com/mmtk/mmtk-core/blob/4c1e58a1b293e0e710feebdb7bec876bd98866f8/src/util/alloc/immix_allocator.rs#L90-L102

Reply:
I remembered @wenyuzhao mentioning something about this in-person to me. Maybe he can chime in and clarify expectations.